packages/nimble-components/src/banner/specs/README.md

```diff
@@ -0,0 +1,186 @@
+
+---
+
+## Design
+
+The banner is a standalone component that spans the width of its containing element, presenting, from left to right, an icon (dependent on Error/Warning/Info type), an optional title, an optional plaintext message, an optional action button or link, and an optional close control. The banner is intended to display short messages (recommended three lines or fewer) that communicate some persistent state or condition.
+
+In this initial implementation, we will not limit the height of the banner. It will grow to fit the text that it is given.
+
+When the user presses the close button, we will hide the banner (and fire a `close` event).
+
+### API
+
+*The key elements of the component's public API surface:*
+
+- *Component Name*: `nimble-banner`
+- *Props/Attrs*:
+  - `heading` - if set, a heading with the specified text is shown to the left of the message. OPEN QUESTION: should we instead use the existing `title` attribute?
+  - `text` - the message text
+  - `prevent-dismiss` - set to hide the close button (attr name taken from Nimble dialog)
+  - `type` - one of `error`, `warning`, `info`, or `neutral` (default)
+  - `action-text` - the text of the action link/button. If unset, no link/button is displayed.
+  - `action-href` - if set, the action element will be a link with the given `href`. (Otherwise, when `action-text` is set, a button element will be displayed.)
+  - `action-button-appearance` - either `outline` or `ghost` (default). When an action button is displayed, this controls the appearance variant.
+  - `close-button-title` - a localized title to give the close button (for a11y purposes)
+- *Methods*
+- *Events*
+  - `close` - fired when the banner is dismissed
+  - `action` - fired when the action button is activated (by mouse or keyboard). This is the intended way for clients to perform an action.
+- *CSS Classes and CSS Custom Properties that affect the component*
+
+### Alternatives
+
+- The text of the banner could be provided as content rather than an attribute. This has the benefit of leaving the door open for us to support more than just plaintext in the future. However, so long as we _don't_ want to support arbitrary HTML, it may have the downside of giving users the wrong idea.
+
+- Instead of configuring the action link/button via attributes, we could allow the user to provide that element as content and slot it in an `action` slot. This would have the following benefits:
+    - allows full configuration of action element (e.g. attributes like `target` on a `nimble-anchor`)
+    - user can add handlers directly to their action element rather than relying on `action` event
+    - easier to support multiple actions in the future
+    - easy to support other kinds of action elements in the future
+
+    However, that flexibility is also a downside. We would no longer have tight control over what the user could show and how it would look.
```

Comment:
I know you mention the flexibility of slotting things as a downside, but I think the current approach has a downside that isn't really captured in the spec, which is that is seems inconsistent with the APIs of all our other components. For pretty much every other nimble component I can think of (button, menu, dialog, tooltip, etc), you slot the configurable pieces into different slots. For the banner, I could envision the following slots:
- title
- content (maybe this is the default slot)
- action-button

I wonder if having tight control over what the user could show is a reason to avoid having a consistent API. I think there are plenty of cases where we "allow" things that designers would frown upon (e.g. putting a hyperlink within the header of a dialog). Do you think this component is significantly different enough from our existing nimble components to have a different API approach?

Reply:
That is a good point. I guess we could also default to hiding elements slotted in the `action` slot and just make exceptions for certain elements (`nimble-button`, `nimble-anchor`), right?

I'm unsure about needing a title slot rather than just an attribute, but maybe it makes sense for completeness/consistency.